### PR TITLE
satellite/satellitedb: ensure that we process orders in order

### PR DESCRIPTION
**Pull request by egonelbre**
What: When transactions are executed in random order concurrently then the postgres database can lockup. Making the transaction fail with a message "database locked".

As a bonus there is a performance improvement:
```
benchmark                                      old ns/op      new ns/op     delta
BenchmarkOrders/Sqlite/Process:50-32           7266013        4679146       -35.60%
BenchmarkOrders/Postgres/Process:50-32         38178964       39148530      +2.54%
BenchmarkOrders/Sqlite#01/Process:100-32       17596919       9265638       -47.35%
BenchmarkOrders/Postgres#01/Process:100-32     79357764       75785671      -4.50%
BenchmarkOrders/Sqlite#02/Process:250-32       70267380       22843365      -67.49%
BenchmarkOrders/Postgres#02/Process:250-32     187335133      177281750     -5.37%
BenchmarkOrders/Sqlite#03/Process:500-32       218870500      45772104      -79.09%
BenchmarkOrders/Postgres#03/Process:500-32     449023067      394666333     -12.11%
BenchmarkOrders/Sqlite#04/Process:999-32       717762000      90001023      -87.46%
BenchmarkOrders/Postgres#04/Process:999-32     1110214300     799247200     -28.01%
```

Why:

Please describe the tests:
 - Test 1:
 - Test 2:
 
Please describe the performance impact:

## Code Review Checklist (to be filled out by reviewer)
 - [x] Does the PR describe what changes are being made?
 - [x] Does the PR describe why the changes are being made?
 - [ ] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [ ] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [x] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [ ] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [ ] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [x] Does any documentation need updating?
 - [x] Do the database access patterns make sense?

### CODE REVIEW

File: satellite/satellitedb/orders.go [hunk 1]
```diff
@@ -236,6 +236,12 @@ func (db *ordersDB) ProcessOrders(ctx context.Context, requests []*orders.Proces
 	if len(requests) == 0 {
 		return nil, err
 	}
+	sort.Slice(requests, func(i, k int) bool {
```
Comment (navillasa) [inline satellite/satellitedb/orders.go]: is this logic tested somewhere already?

**Comment by stefanbenten** [comment]
We could use a test to ensure this type of error is never happening again :)
Maybe @ethanadams has an easy test to add?

File: pkg/storj/serialnumber.go [hunk 2]
```diff
@@ -44,6 +44,18 @@ func (id SerialNumber) IsZero() bool {
 	return id == SerialNumber{}
 }
 
+// Less returns whether id is smaller than b in lexiographic order
+func (id SerialNumber) Less(b SerialNumber) bool {
```
Comment (navillasa) [inline pkg/storj/serialnumber.go]: nit - why is b the name for the SerialNumber argument?

Comment (egonelbre) [reply pkg/storj/serialnumber.go]: Because I couldn't figure out a better name.

File: satellite/satellitedb/orders.go (hunk 1 above)
Comment (egonelbre) [reply satellite/satellitedb/orders.go]: Not really, but @ethanadams has a reproducer.

Comment (ethanadams) [reply satellite/satellitedb/orders.go]: @navillasa are you referring to testing the sort logic, or testing the fact that this change fixes the deadlock issue?

**Comment by littleskunk** [comment]
Negative. Deadlock still in place.
```
satellite/peer.go:722	Node 18zEcpEtGjwtXT9tBbJnKPfyGC3P3ZbwFZkzEyhiYajtmzRBYS started
[...]
storagenode/1    12gPTSsJdz 21:55:22.124 | ERROR	orders.18zEcpEtGjwtXT9tBbJnKPfyGC3P3ZbwFZkzEyhiYajtmzRBYS	orders/service.go:194	failed to settle orders	{"error": "order: failed to receive settlement response: rpc error: code = Unknown desc = pq: deadlock detected", "errorVerbose": "order: failed to receive settlement response: rpc error: code = Unknown desc = pq: deadlock detected\n\tstorj.io/storj/storagenode/orders.(*Service).settle:312\n\tstorj.io/storj/storagenode/orders.(*Service).Settle:192\n\tstorj.io/storj/storagenode/orders.(*Service).sendOrders.func2:171\n\tgolang.org/x/sync/errgroup.(*Group).Go.func1:57"}
                                         | storj.io/storj/storagenode/orders.(*Service).Settle
                                         | 	/home/newrelease/storj/storagenode/orders/service.go:194
                                         | storj.io/storj/storagenode/orders.(*Service).sendOrders.func2
                                         | 	/home/newrelease/storj/storagenode/orders/service.go:171
                                         | golang.org/x/sync/errgroup.(*Group).Go.func1
                                         | 	/home/newrelease/go/pkg/mod/golang.org/x/sync@v0.0.0-20190423024810-112230192c58/errgroup/errgroup.go:57
```

File: satellite/satellitedb/orders.go (hunk 1 above)
Comment (navillasa) [reply satellite/satellitedb/orders.go]: I think I meant testing the sort logic, but maybe testing that it fixes the deadlock is more important.

Comment (ethanadams) [reply satellite/satellitedb/orders.go]: Apparently my test was not comprehensive enough.  I was uploading a lot of small files and then restarting the sim.  This did reproduce the issue before the sorting.  I received some help from @littleskunk and have updated my script to this:
```
$ for (( c=10081; c<=10100; c++ )) do uplink --config-dir "/Users/ethan/Library/Application Support/Storj/masterpg" cp ~/Downloads/< 70M file > sj://test3/$c; uplink --config-dir "/Users/ethan/Library/Application Support/Storj/masterpg" cp sj://test3/$c /dev/null; done
```
I can now reproduce.

File: pkg/storj/serialnumber.go (hunk 2 above)
Comment (ifraixedes) [reply pkg/storj/serialnumber.go]: not the best name, but `another` or `other` looks better to me.

Comment (egonelbre) [reply pkg/storj/serialnumber.go]: used `other`

File: pkg/storj/serialnumber.go [hunk 3]
```diff
@@ -44,6 +44,18 @@ func (id SerialNumber) IsZero() bool {
 	return id == SerialNumber{}
 }
 
+// Less returns whether id is smaller than other in lexicographic order
```
Comment (ifraixedes) [inline pkg/storj/serialnumber.go]:
```suggestion
// Less returns whether id is smaller than other in lexicographic order.
```

File: pkg/storj/node.go [hunk 4]
```diff
@@ -110,12 +110,12 @@ func (id NodeID) IsZero() bool {
 // Bytes returns raw bytes of the id
 func (id NodeID) Bytes() []byte { return id[:] }
 
-// Less returns whether id is smaller than b in lexiographic order
-func (id NodeID) Less(b NodeID) bool {
+// Less returns whether id is smaller than other in lexicographic order
```
Comment (ifraixedes) [inline pkg/storj/node.go]:
```suggestion
// Less returns whether id is smaller than other in lexicographic order.
```

File: satellite/satellitedb/orders.go [hunk 5]
```diff
@@ -237,6 +235,19 @@ func (db *ordersDB) ProcessOrders(ctx context.Context, requests []*orders.Proces
 		return nil, err
 	}
 
+	// check that all requests are from the same storage node
+	storageNodeID := requests[0].OrderLimit.StorageNodeId
+	for _, req := range requests[1:] {
+		if req.OrderLimit.StorageNodeId != storageNodeID {
+			return nil, Error.New("requests from different different storage nodes %v and %v", storageNodeID, req.OrderLimit.StorageNodeId)
```
Comment (ifraixedes) [inline satellite/satellitedb/orders.go]:
```suggestion
			return nil, Error.New("requests from different storage nodes %v and %v", storageNodeID, req.OrderLimit.StorageNodeId)
```

File: satellite/satellitedb/orders.go [hunk 6]
```diff
@@ -237,6 +235,19 @@ func (db *ordersDB) ProcessOrders(ctx context.Context, requests []*orders.Proces
 		return nil, err
 	}
 
+	// check that all requests are from the same storage node
```
Comment (ifraixedes) [inline satellite/satellitedb/orders.go]: nit: I would add to the function doc comment about this restriction.

File: satellite/satellitedb/orders.go [hunk 7]
```diff
@@ -229,14 +227,29 @@ func (db *ordersDB) UnuseSerialNumber(ctx context.Context, serialNumber storj.Se
...
 
 	if len(requests) == 0 {
 		return nil, err
 	}
 
+	// check that all requests are from the same storage node
+	storageNodeID := requests[0].OrderLimit.StorageNodeId
+	for _, req := range requests[1:] {
+		if req.OrderLimit.StorageNodeId != storageNodeID {
+			return nil, Error.New("requests from different storage nodes %v and %v", storageNodeID, req.OrderLimit.StorageNodeId)
```
Comment (jtolds) [inline satellite/satellitedb/orders.go]: new error class?

Comment (jtolds) [reply satellite/satellitedb/orders.go]: var (
  ErrDuplicateStorageNode = errs.Class("duplicate storage node")
)

ErrDuplicateStorageNode.New("...")

**Pull request merged**